src/run_program.rs

```diff
@@ -347,9 +493,11 @@ impl<'a, D: Dialect> RunProgramContext<'a, D> {
                 None => break,
             };
             cost += match op {
-                Operation::Apply => {
-                    augment_cost_errors(self.apply_op(max_cost - cost), max_cost_ptr)?
-                }
+                Operation::Apply => augment_cost_errors(
+                    self.apply_op(cost, effective_max_cost - cost),
+                    max_cost_ptr,
+                )?,
+                Operation::Softfork => self.exit_softfork_guard(cost)?,
```

Comment:
The names are a bit inconsistent here. Maybe rename the operator to match the name of the function it calls? `exit_guard` & `ExitGuard`, or `EndSoftfork` and `end_soft_fork`, or whatever.